controller/lib/core/nvparam.cpp

```diff
@@ -0,0 +1,218 @@
+static bool CheckStruct(uint32_t addr);
+static uint32_t BlockCRC(void *blk);
+static void Invalidate(uint32_t addr);
+static bool SaveBlock(NVparams *blk, uint32_t addr);
+
+// One time init of non-volatile parameter area.
+void NVparamsInit(void) {
+  // Make sure the size of the storage structure is correct.
+  // This catches mistakes when adding new variables.
+  // If it's wrong, then I call a function that doesn't actually
+  // exist.  That generates a link time error, so the program
+  // won't actually be built.
+  // If the size is correct then the compiler will optimize this out.
+  if (sizeof(NVparams) != nvparam_size) {
+    extern void CheckSizeOfNVparams(void);
+    CheckSizeOfNVparams();
+  }
+
+  // Total structures in two pages of flash
+  int N = flash_params_size / nvparam_size;
+
+  // Run through structures looking for a valid one.
+  uint32_t addr = flash_params_start;
+  for (int i = 0; i < N; i++, addr += nvparam_size) {
+    // Check the structure at this address
+    if (!CheckStruct(addr))
+      continue;
+
+    // If this is the first valid block we've found,
+    // just keep track of the address
+    if (!nvparam_addr) {
+      nvparam_addr = addr;
+      continue;
+    }
+
+    // I've found multiple valid blocks.  This normally shouldn't
+    // happen but could if power is lost at just the right time
+    // when updating flash.
+    //
+    // I'll check the counter to see which was is newer and keep
+    // that address.  I'll also invalidate the older one.
+    NVparams *a = reinterpret_cast<NVparams *>(nvparam_addr);
+    NVparams *b = reinterpret_cast<NVparams *>(addr);
+
+    int8_t diff = static_cast<int8_t>(b->count - a->count);
+    if (diff > 0) {
+      Invalidate(nvparam_addr);
+      nvparam_addr = addr;
+    } else
+      Invalidate(addr);
+  }
+
+  // If I found a valid parameter area, I'm done
+  if (nvparam_addr)
+    return;
+
+  // If no valid storage was found then I'll just create one with
+  // zero values.
+  NVparams blank;
+  memset(&blank, 0, sizeof(blank));
+  blank.mark = good_mark;
+  blank.crc = BlockCRC(&blank);
+
+  Hal.FlashErasePage(flash_params_start);
+  Hal.FlashWrite(flash_params_start, reinterpret_cast<uint32_t *>(&blank),
+                 sizeof(blank) / sizeof(uint32_t));
+  nvparam_addr = flash_params_start;
+}
+
+const NVparams *FindStore(void) {
+  return reinterpret_cast<const NVparams *>(nvparam_addr);
+}
+
+bool NVparamsUpdtOff(uint32_t offset, const void *value, uint8_t len) {
+
+  // Make sure the passed pointer is pointing to somewhere
+  // in the current page and isn't in the reserved first 8 bytes
+  if ((offset < 8) || (offset >= nvparam_size))
+    return false;
+
+  // Copy the current parameter block to RAM
+  NVparams temp;
+  memcpy(&temp, reinterpret_cast<void *>(nvparam_addr), nvparam_size);
+
+  // Update the contents in RAM
+  memcpy(reinterpret_cast<uint8_t *>(&temp) + offset, value, len);
+
+  // Make sure the reserved area is kept at zero
```

Comment:
This is indeed important for the CRC.

It's also important that there be no padding between the end of the header and the beginning of the rsvd space, otherwise when we calculate the CRC, it will include the padding bits, whose values are undefined.  So I think you should make rsvd a char[] instead of uint32[].